contracts/governance/utils/VotesConfidential.sol

```diff
@@ -0,0 +1,209 @@
+     * by the derived contract.
+     */
+    function totalSupply() public view virtual returns (euint64);
+
+    /// @dev Returns the current total supply of votes.
+    function getCurrentTotalSupply() public view virtual returns (euint64) {
+        return _totalCheckpoints.latest();
+    }
+
+    /// @dev Returns the delegate that `account` has chosen.
+    function delegates(address account) public view virtual returns (address) {
+        return _delegatee[account];
+    }
+
+    /// @dev Delegates votes from the sender to `delegatee`.
+    function delegate(address delegatee) public virtual {
+        _delegate(msg.sender, delegatee);
+    }
+
+    /// @dev Delegates votes from `delegator` to `delegatee`.
+    function delegateBySig(
+        address delegator,
+        address delegatee,
+        uint256 nonce,
+        uint256 expiry,
+        bytes memory signature
+    ) public virtual {
+        if (block.timestamp > expiry) {
+            revert VotesExpiredSignature(expiry);
+        }
+        if (
+            !SignatureChecker.isValidSignatureNow(
+                delegator,
+                _hashTypedDataV4(keccak256(abi.encode(DELEGATION_TYPEHASH, delegatee, nonce, expiry))),
+                signature
+            )
+        ) {
+            revert VotesInvalidSignature(delegator);
+        }
+        _useCheckedNonce(delegator, nonce);
+        _delegate(delegator, delegatee);
+    }
+
+    /**
+     * @dev Delegate all of `account`'s voting units to `delegatee`.
+     *
+     * Emits events {IVotes-DelegateChanged} and {IVotes-DelegateVotesChanged}.
+     */
+    function _delegate(address account, address delegatee) internal virtual {
+        address oldDelegate = delegates(account);
+        _delegatee[account] = delegatee;
+
+        emit DelegateChanged(account, oldDelegate, delegatee);
+        _moveDelegateVotes(oldDelegate, delegatee, _getVotingUnits(account));
+    }
+
+    /**
+     * @dev Transfers, mints, or burns voting units. To register a mint, `from` should be zero. To register a burn, `to`
+     * should be zero. Total supply of voting units will be adjusted with mints and burns.
+     */
+    function _transferVotingUnits(address from, address to, euint64 amount) internal virtual {
+        if (from == address(0) || to == address(0)) {
+            _push(_totalCheckpoints, totalSupply());
+        }
+        _moveDelegateVotes(delegates(from), delegates(to), amount);
+    }
+
+    /**
+     * @dev Moves delegated votes from one delegate to another.
+     */
+    function _moveDelegateVotes(address from, address to, euint64 amount) internal virtual {
+        CheckpointsConfidential.TraceEuint64 storage store;
+        if (from != to && euint64.unwrap(amount) != 0) {
+            if (from != address(0)) {
+                store = _delegateCheckpoints[from];
+                euint64 newValue = store.latest().sub(amount);
```

Comment:
`store.latest()` might be 0 as in "was never created/allocated". These should have special treatment

I believe https://github.com/OpenZeppelin/openzeppelin-confidential-contracts/blob/master/contracts/utils/TFHESafeMath.sol#L16-L25 would be usefull here.

Reply:
That makes me think the Votes.sol construction, with a `_push` function that takes a fnPointer, would probably work fine with the functions in THFESafeMath.sol